app/services/oidc_jwt_validate.rb

```diff
@@ -0,0 +1,50 @@
+class OidcJwtValidate
+  attr_reader :decoded_token
+  delegate :logger, to: :Rails
+
+  def initialize(jwt_token)
+    @jwt_token = jwt_token
+  end
+
+  def decoded_payload
+    # OpenSSL#set_key method does not support ruby version < 2.4.0, apparently the JWT gem uses
+    # OpenSSL#set_key method for all ruby version. We must remove this condition once new version
+    # of the JWT(2.2.2) is released.
+    unless OpenSSL::PKey::RSA.new.respond_to?(:set_key)
+      Foreman::Logging.logger('app').error "SSO feature is not available for Ruby < 2.4.0"
+      return nil
+    end
+    JWT.decode(@jwt_token, nil, true,
+      { aud: Setting['oidc_audience'],
+        verify_aud: true,
+        iss: Setting['oidc_issuer'],
+        verify_iss: true,
+        algorithms: [Setting['oidc_algorithm']],
+        jwks: jwks_loader }
+    ).first
+  rescue JWT::DecodeError => e
+    Foreman::Logging.exception('Failed to decode JWT', e)
+    nil
+  end
+
+  private
+
+  def jwks_loader(options = {})
+    response = RestClient::Request.execute(
+      :url => Setting['oidc_jwks_url'],
+      :method => :get,
+      :verify_ssl => true
+    )
+    json_response = JSON.parse(response)
+    if json_response.is_a?(Hash)
+      jwks_keys = json_response['keys']
+      @cached_keys = nil if options[:invalidate] # need to reload the keys
+      @cached_keys ||= { keys: jwks_keys.map(&:symbolize_keys) }
```

Comment:
doing the cache check here beats the whole point of using a cache - i.e. to save the potentially slow network request. It should be checked before the request is made. 
However, thinking about it again, I don't think there is much to gain by caching this at the instance level - since we are creating a new instance of the validator on every request. I'd be fine with skipping the caching at first until we see if it is worth the effort to do properly (but leave the `(options = {})` parameter so we are compatible with the function fingerprint JWT expects)